File: BakhoumEtAl2017/RNAseq_Samuel4427_2017_02_08/PGE/main.R
library(circlize)
library(xlsx)

plot_data <- function(prefix,de_genes,pge.results) {

  gene_pos <- read.csv("ensembl89.csv",row.names=1,header=F,colClasses=c("character","character","numeric","numeric","character"))
  gene_pos <- gene_pos[row.names(de_genes),]

  colors <- rep("red",length(de_genes$log2FoldChange))
  colors[de_genes$log2FoldChange<0]<-"blue"

  de_data <- data.frame(
    chr=gene_pos[,1],
    start=gene_pos[,2],
    end=gene_pos[,3],
    value1=de_genes$log2FoldChange,
    color=rep("red",length(de_genes$log2FoldChange)),
    stringsAsFactors=FALSE
  )
  de_data[de_data$value1<0,"color"]<-"blue"
  de_data$chr <- unlist(lapply(de_data$chr,function(x) return(paste("chr",x,sep=""))))
  de_data <- de_data[(de_data$chr!="chrMT") & (de_data$chr!="chrNA"),]


  pdf(paste(prefix,"circos.pdf",sep=""))
  circos.par("gap.degree"=c(rep(1,23),7),"start.degree"=90)
  circos.initializeWithIdeogram(
    cytoband="/Users/charlesmurphy/Desktop/Research/data/UCSC/hg38/cytoBand-canonical.txt"
  )
  circos.genomicTrackPlotRegion(
    de_data[,c("chr","start","end","value1")]
  )
  circos.yaxis(labels.cex = 0.6,sector.index="chr1")
  for (chrom in unique(de_data$chr)) {
    tmp <- de_data[de_data$chr==chrom,]
    circos.genomicPoints(region=tmp[,c("start","end")],value=data.frame(value=tmp$value1),col=tmp$color,sector.index=chrom,pch=".",cex=2.5)
  }
  circos.genomicTrackPlotRegion(
    data.frame(
      chr=pge.results$chrom,
      start=pge.results$start,
      end=pge.results$end,
      stringsAsFactors=FALSE
    ),
    ylim=c(0,max(-log10(pge.results$padj))*1.1)
  )
  circos.yaxis(labels.cex = 0.6,sector.index="chr1")
  for (chrom in unique(pge.results$chrom)) {
    tmp <- pge.results[pge.results$chrom==chrom,]
    colors <- as.character(tmp$cna)
    colors[colors=="amplification"] <- "red"
    colors[colors=="deletion"] <- "blue"
    circos.genomicLines(
      region=tmp[,c("start","end")],
      value=data.frame(value=-log10(tmp$padj)),
      type = "segment",
      col=colors,
      sector.index=chrom,
      lwd = 4
    )
  }
  dev.off()
  circos.clear()
}

comparisons <- c('BoneMet_vs_BrainMet')
padj <- c(0.05,0.1,0.25,1.0)
log2s <- c(0,0.25,0.5)
pvals <- c(0.05)

for (comparison in comparisons) {
  for (pv in pvals) {
    for (pp in padj) {
      for (ll in log2s) {

        de_genes <- read.csv(
          paste("/Users/charlesmurphy/Desktop/Research/0816_sam/results/RNAseq_Samuel4427_2017_02_08/DESeq2_ensembl/",comparison,"/",comparison,"_results.csv",sep=""),
          header=T,
          check.names=F,
          row.names=1,
          colClasses=c("character","numeric","numeric","numeric","numeric","numeric","numeric")
        )
        de_genes <- de_genes[(!is.na(de_genes$padj)) & (de_genes$padj<=pp) & (de_genes$pvalue<=pv) & (abs(de_genes$log2FoldChange)>=ll),]

        if (!file.exists(paste("./",comparison,"/pval",pv,"_padj",pp,"_log2FC",ll,"/",comparison,"_results.xlsx",sep=""))) {
          next
        }

        pge.results <- read.xlsx(
          paste("./",comparison,"/pval",pv,"_padj",pp,"_log2FC",ll,"/",comparison,"_results.xlsx",sep=""),
          "Sheet1",
          colClasses=c("character",rep("numeric",4),"character","numeric","numeric","character")
        )
        pge.results <- pge.results[(pge.results$num_differentially_expressed_genes>3) & (pge.results$padj<=0.01),]
        pge.results$chrom <- unlist(lapply(pge.results$chrom,function(x) return(paste("chr",x,sep=""))))

        plot_data(
          paste("./",comparison,"/pval",pv,"_padj",pp,"_log2FC",ll,"/",sep=""),
          de_genes,
          pge.results
        )
      }
    }
  }
}

File: BakhoumEtAl2017/RNAseq_Samuel4427_2017_02_08/PGE/main.py
import os
import pandas
import pyensembl

def parse_result_file(filename):
    mapping = []
    raw = []
    bed = []
    index=0


    data = open(filename,'r').read().split('\n')
    mapping.append(data[index].replace('<mapping>','').split(','))
    index+=1
    while data[index].find('</mapping><raw>')==-1:
        mapping.append(data[index].split(','))
        index+=1
    if data[index].find('</mapping><raw></raw>')==-1:
        raw.append(data[index].replace('</mapping><raw>','').split('\t'))
        index+=1
        while data[index].find('</raw><bed>')==-1:
            raw.append(data[index].split('\t'))
            index+=1
    else:
        index+=1

    while (index < len(data)) and (data[index].find('</bed>')==-1):
        bed.append(data[index].split('\t'))
        index+=1

    mapping = pandas.DataFrame(mapping,columns=['gene_id1','gene_id2','chrom','start','end','gene_id3'])
    raw = pandas.DataFrame(raw,columns=['chrom','start','end','pvalue','padj','num_differentially_expressed_genes','total_genes_in_region'])
    bed = pandas.DataFrame(bed,columns=['chrom','start','end','pval'])

    return mapping,raw,bed


def prep_data(comparisons,comparisons_params):
    for comparison in comparisons:
        outdir = comparison + '/pval' + str(comparisons_params[comparison]['pvalue']) + '_padj' + str(comparisons_params[comparison]['padj']) + '_log2FC' + str(comparisons_params[comparison]['log2'])
        if not os.path.exists(outdir):
            os.mkdir(outdir)
        deseq = pandas.read_csv("/Users/charlesmurphy/Desktop/Research/0816_sam/results/RNAseq_Samuel4427_2017_02_08/DESeq2_ensembl/" + comparison + "/" + comparison + "_results.csv")

        deseq2 = deseq[(deseq['padj']<=comparisons_params[comparison]['padj']) & (deseq['log2FoldChange'] >= comparisons_params[comparison]['log2']) & (deseq['pvalue']<=comparisons_params[comparison]['pvalue'])]
        deseq2['Unnamed: 0'].to_csv(outdir + '/' + comparison + '_up.txt',index=False,sep='\t')

        deseq2 = deseq[(deseq['padj']<=comparisons_params[comparison]['padj']) & (deseq['log2FoldChange']<= -comparisons_params[comparison]['log2']) & (deseq['pvalue']<=comparisons_params[comparison]['pvalue'])]
        deseq2['Unnamed: 0'].to_csv(outdir + '/' + comparison + '_down.txt',index=False,sep='\t')

def getOverlap(a, b):
    return max(0, min(a[1], b[1]) - max(a[0], b[0]))

def get_overlapping_genes(raw,mapping,genes):
    overlapping_genes = []
    mapping['chrom']=mapping['chrom'].str.upper()
    for i in raw.index:
        region_start = int(raw.loc[i,'start'])
        region_end = int(raw.loc[i,'end'])
        chrom = raw.loc[i,'chrom']
        mapping_sub = mapping[mapping['chrom']==chrom]

        subset = []
        for j in mapping_sub.index:
            j_start = int(mapping_sub.loc[j,'start'])
            j_end = int(mapping_sub.loc[j,'end'])
            if getOverlap([region_start,region_end],[j_start,j_end])>0:
                subset.append(True)
            else:
                subset.append(False)

        mapping_sub = mapping_sub[subset]['gene_id1'].tolist()
        mapping_sub = map(lambda x: x.upper(),mapping_sub)
        mapping_sub = ','.join(genes.loc[mapping_sub,'symbols'].tolist())
        overlapping_genes.append(mapping_sub)

    raw['differentially_expressed_genes'] = overlapping_genes
    raw['padj'] = raw['padj'].astype(float)
    raw = raw.sort_values(['padj'])
    return raw

def process_results(comparisons,comparisons_params):

    db = pyensembl.EnsemblRelease('75','human')

    for directory in comparisons:
        outdir = directory + '/pval' + str(comparisons_params[directory]['pvalue']) + '_padj' + str(comparisons_params[directory]['padj']) + '_log2FC' + str(comparisons_params[directory]['log2'])
        print outdir

        try:
            genes = pandas.read_table('./' + outdir + '/' + directory + '_up.txt',header=None)
        except:
            continue

        symbols = []
        for ii in genes[0].tolist():
            try:
                symbols.append(db.gene_by_id(ii).gene_name)
            except ValueError:
                symbols.append(ii)
        genes['symbols'] = symbols
        genes.index = genes[0]

        mapping,raw,bed = parse_result_file(
            './' + outdir + '/' + directory + '_results_up.txt'
        )
        raw = get_overlapping_genes(raw,mapping,genes)
        raw_up = raw[raw['total_genes_in_region'].astype(int)>3]
        raw_up['cna'] = ['amplification']*raw_up.shape[0]

        try:
            genes = pandas.read_table('./' + outdir + '/' + directory + '_down.txt',header=None)
        except:
            continue
        symbols = []
        for ii in genes[0].tolist():
            try:
                symbols.append(db.gene_by_id(ii).gene_name)
            except ValueError:
                symbols.append(ii)
        genes['symbols'] = symbols
        genes.index = genes[0]

        mapping,raw,bed = parse_result_file(
            './' + outdir + '/' + directory + '_results_down.txt'
        )
        raw = get_overlapping_genes(raw,mapping,genes)
        raw_down = raw[raw['total_genes_in_region'].astype(int)>3]
        raw_down['cna'] = ['deletion']*raw_down.shape[0]

        raw = raw_down.append(raw_up)
        raw = raw[['chrom','start','end','pvalue','padj','cna','num_differentially_expressed_genes','total_genes_in_region','differentially_expressed_genes']]
        raw = raw.sort_values('padj')

        raw.to_excel('./' + outdir + '/' + directory + '_results.xlsx',index=False)

def run_data(fout,comparisons,comparisons_params):
    for comparison in comparisons:
        outdir = comparison + '/pval' + str(comparisons_params[comparison]['pvalue']) + '_padj' + str(comparisons_params[comparison]['padj']) + '_log2FC' + str(comparisons_params[comparison]['log2'])
        fout.write(
            'perl pge.pl -q ./%s/%s_up.txt -r ensembl89 > ./%s/%s_results_up.txt\n' %
            (
                outdir,
                comparison,
                outdir,
                comparison
            )
        )
        fout.write(
            'perl pge.pl -q ./%s/%s_down.txt -r ensembl89 > ./%s/%s_results_down.txt\n' %
            (
                outdir,
                comparison,
                outdir,
                comparison
            )
        )

comparisons = [
    'primary_vs_metastasis','BoneMet_vs_BrainMet',
    'CINhighPrimary_vs_CINhighMetastasis','CINLowMets_vs_CINHighMets',
    'CINLowPrimary_vs_CINMedHighPrimary','Control_vs_STINGshRNA',
    'CINlow_vs_CINmedhigh']
comparisons=['1','2','3','23','24','31','32','45']


fout = open('pge_scripts.sh','w')
for pv in [0.05]:
    #for log2 in [0,0.25,0.5]:
    for log2 in [0]:
        #for padj in [0.05,0.1,0.25,1]:
        for padj in [0.05]:
            comparisons_params = {}
            comparisons_params['1'] = {'padj':padj,'pvalue':pv,'log2':log2}
            comparisons_params['2'] = {'padj':padj,'pvalue':pv,'log2':log2}
            comparisons_params['3'] = {'padj':padj,'pvalue':pv,'log2':log2}
            comparisons_params['23'] = {'padj':padj,'pvalue':pv,'log2':log2}
            comparisons_params['24'] = {'padj':padj,'pvalue':pv,'log2':log2}
            comparisons_params['31'] = {'padj':padj,'pvalue':pv,'log2':log2}
            comparisons_params['32'] = {'padj':padj,'pvalue':pv,'log2':log2}
            comparisons_params['45'] = {'padj':padj,'pvalue':pv,'log2':log2}
            #comparisons_params['primary_vs_metastasis'] = {'padj':padj,'pvalue':pv,'log2':log2}
            #comparisons_params['BoneMet_vs_BrainMet'] = {'padj':padj,'pvalue':pv,'log2':log2}
            #comparisons_params['CINhighPrimary_vs_CINhighMetastasis'] = {'padj':padj,'pvalue':pv,'log2':log2}
            #comparisons_params['CINLowMets_vs_CINHighMets'] = {'padj':padj,'pvalue':pv,'log2':log2}
            #comparisons_params['CINLowPrimary_vs_CINMedHighPrimary'] = {'padj':padj,'pvalue':pv,'log2':log2}
            #comparisons_params['Control_vs_STINGshRNA'] = {'padj':padj,'pvalue':pv,'log2':log2}
            #comparisons_params['CINlow_vs_CINmedhigh'] = {'padj':padj,'pvalue':pv,'log2':log2}

            #prep_data(comparisons,comparisons_params)
            #run_data(fout,comparisons,comparisons_params)
fout.close()
#exit()
for pv in [0.05]:
    #for log2 in [0,0.25,0.5]:
    for log2 in [0]:
        #for padj in [0.05,0.1,0.25,1]:
        for padj in [0.05]:
            comparisons_params = {}
            comparisons_params['1'] = {'padj':padj,'pvalue':pv,'log2':log2}
            comparisons_params['2'] = {'padj':padj,'pvalue':pv,'log2':log2}
            comparisons_params['3'] = {'padj':padj,'pvalue':pv,'log2':log2}
            comparisons_params['23'] = {'padj':padj,'pvalue':pv,'log2':log2}
            comparisons_params['24'] = {'padj':padj,'pvalue':pv,'log2':log2}
            comparisons_params['31'] = {'padj':padj,'pvalue':pv,'log2':log2}
            comparisons_params['32'] = {'padj':padj,'pvalue':pv,'log2':log2}
            comparisons_params['45'] = {'padj':padj,'pvalue':pv,'log2':log2}
            #comparisons_params['primary_vs_metastasis'] = {'padj':padj,'pvalue':pv,'log2':log2}
            #comparisons_params['BoneMet_vs_BrainMet'] = {'padj':padj,'pvalue':pv,'log2':log2}
            #comparisons_params['CINhighPrimary_vs_CINhighMetastasis'] = {'padj':padj,'pvalue':pv,'log2':log2}
            #comparisons_params['CINLowMets_vs_CINHighMets'] = {'padj':padj,'pvalue':pv,'log2':log2}
            #comparisons_params['CINLowPrimary_vs_CINMedHighPrimary'] = {'padj':padj,'pvalue':pv,'log2':log2}
            #comparisons_params['Control_vs_STINGshRNA'] = {'padj':padj,'pvalue':pv,'log2':log2}
            #comparisons_params['CINlow_vs_CINmedhigh'] = {'padj':padj,'pvalue':pv,'log2':log2}

            process_results(comparisons,comparisons_params)
